host_applications/linux/apps/raspicam/RaspiGPS.c

```diff
@@ -0,0 +1,188 @@
+
+void raspi_gps_unlock()
+{
+   pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+}
+
+
+static void *gps_reader_process(void *gps_reader_data_ptr)
+{
+   while (!gps_reader_data.terminated)
+   {
+      int ret = 0;
+      int gps_valid = 0;
+
+      gps_reader_data.gpsd.gpsdata.set = 0;
+      gps_reader_data.gpsd.gpsdata.fix.mode = 0;
+      if ((connect_gpsd(&gps_reader_data.gpsd) < 0) ||
+            ((ret = read_gps_data_once(&gps_reader_data.gpsd)) < 0))
+         break;
+
+      if (ret > 0 && gps_reader_data.gpsd.gpsdata.online)
+      {
+         if (gps_reader_data.gpsd.gpsdata.fix.mode >= MODE_2D)
+         {
+            // we have GPS fix, copy fresh data to cache
+            gps_valid = 1;
+            time(&gps_reader_data.last_valid_time);
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            memcpy(&gps_reader_data.gpsdata_cache, &gps_reader_data.gpsd.gpsdata,
+                   sizeof(struct gps_data_t));
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+      }
+      if (!gps_valid)
+      {
+         time_t now;
+         time(&now);
+         if (now - gps_reader_data.last_valid_time > GPS_CACHE_EXPIRY)
+         {
+            // our cache is stale, clear it
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            gps_reader_data.gpsdata_cache.online = gps_reader_data.gpsd.gpsdata.online;
+            gps_reader_data.gpsdata_cache.set = 0;
+            gps_reader_data.gpsdata_cache.fix.mode = 0;
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+         // we lost GPS fix, copy GPS time to cache if available
+         if (gps_reader_data.gpsd.gpsdata.set & TIME_SET)
+         {
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            gps_reader_data.gpsdata_cache.set |= TIME_SET;
+            gps_reader_data.gpsdata_cache.fix.time = gps_reader_data.gpsd.gpsdata.fix.time;
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+      }
+   }
+   return NULL;
+}
+
+void raspi_gps_shutdown(int verbose)
+{
+   gps_reader_data.terminated = 1;
+
+   if (gps_reader_data.gps_reader_thread_ok)
+   {
+      if (verbose)
+         fprintf(stderr, "Waiting for GPS reader thread to terminate\n");
+
+      pthread_join(gps_reader_data.gps_reader_thread, NULL);
+   }
+   if (verbose && gps_reader_data.gpsd.gpsd_connected)
+      fprintf(stderr, "Closing gpsd connection\n\n");
+
+   disconnect_gpsd(&gps_reader_data.gpsd);
+
+   libgps_unload(&gps_reader_data.gpsd);
+
+   pthread_mutex_destroy(&gps_reader_data.gps_cache_mutex);
+}
+
+
+void raspi_gps_setup(int verbose)
+{
+   memset(&gps_reader_data, 0, sizeof(gps_reader_data));
+
+   pthread_mutex_init(&gps_reader_data.gps_cache_mutex, NULL);
+
+   gpsd_init(&gps_reader_data.gpsd);
+   if (libgps_load(&gps_reader_data.gpsd))
+   {
+      pthread_mutex_destroy(&gps_reader_data.gps_cache_mutex);
+      exit(EX_SOFTWARE);
```

Comment:
exit here seems a tad extreme. Return error and exit from the caller (if that is deemed necessary). Ditto the other failure cases in this function.

Reply:
Hmm, the original dropped out, not nececessarily so abruptly. Will ponder how to make this neater.

Reply:
Cleanup (not needed here as gpsd_init only does a memset and writes a couple of values) and return an error code.
The caller can abort on error, but it makes this a more useful library than having it randomly abort the app.

The code looks to be doing all the cleanup already, so simply replacing the ```exit``` calls with return -1, and return 0 at the end would do it. (Function needs to change from returning void to returning int too).